pennylane_qulacs/qulacs_device.py

```diff
@@ -39,181 +74,239 @@
     """Qulacs device"""
-    name = 'Qulacs device'
-    short_name = 'qulacs.simulator'
-    pennylane_requires = '>=0.5.0'
+    name = "Qulacs device"
+    short_name = "qulacs.simulator"
+    pennylane_requires = ">=0.10.0"
     version = __version__
-    author = 'Steven Oud and Xanadu'
+    author = "Steven Oud and Xanadu"
+    gpu_supported = GPU_SUPPORTED
 
     _capabilities = {
-        'model': 'qubit',
-        'tensor_observables': True
+        "model": "qubit",
+        "tensor_observables": True,
+        "inverse_operations": True
     }
 
-    _operations_map = {
-        'QubitStateVector': None,
-        'BasisState': None,
-        'QubitUnitary': None,
-        'Toffoli': toffoli,
-        'CSWAP': CSWAP,
-        'CRZ': crz,
-        'Rot': None,
-        'SWAP': gate.SWAP,
-        'CNOT': gate.CNOT,
-        'CZ': gate.CZ,
-        'S': gate.S,
-        'Sdg': gate.Sdag,
-        'T': gate.T,
-        'Tdg': gate.Tdag,
-        'RX': gate.RX,
-        'RY': gate.RY,
-        'RZ': gate.RZ,
-        'PauliX': gate.X,
-        'PauliY': gate.Y,
-        'PauliZ': gate.Z,
-        'Hadamard': gate.H
+    _operation_map = {
+        "QubitStateVector": None,
+        "BasisState": None,
+        "QubitUnitary": None,
+        "Toffoli": gate.TOFFOLI,
+        "CSWAP": gate.FREDKIN,
+        "CRZ": crz,
+        "SWAP": gate.SWAP,
+        "CNOT": gate.CNOT,
+        "CZ": gate.CZ,
+        "S": gate.S,
+        "T": gate.T,
+        "RX": gate.RX,
+        "RY": gate.RY,
+        "RZ": gate.RZ,
+        "Rot": None,
+        "PauliX": gate.X,
+        "PauliY": gate.Y,
+        "PauliZ": gate.Z,
+        "Hadamard": gate.H,
+        "PhaseShift": phase_shift
     }
+
     _observable_map = {
-        'PauliX': X,
-        'PauliY': Y,
-        'PauliZ': Z,
-        'Hadamard': H,
-        'Identity': I,
-        'Hermitian': hermitian
+        "PauliX": X,
+        "PauliY": Y,
+        "PauliZ": Z,
+        "Hadamard": H,
+        "Identity": I,
+        "Hermitian": None
     }
 
-    operations = _operations_map.keys()
+    operations = _operation_map.keys()
     observables = _observable_map.keys()
 
-    def __init__(self, wires, gpu=False, **kwargs):
-        super().__init__(wires=wires)
+    # Add inverse gates to _operation_map
+    _operation_map.update({k + ".inv": v for k, v in _operation_map.items()})
+
+    def __init__(self, wires, shots=1000, analytic=True, gpu=False, **kwargs):
+        super().__init__(wires=wires, shots=shots, analytic=analytic)
 
         if gpu:
-            if not GPU_SUPPORTED:
+            if not QulacsDevice.gpu_supported:
                 raise DeviceError(
-                    'GPU not supported with installed version of qulacs. '
-                    'Please install "qulacs-gpu" to use GPU simulation.'
+                    "GPU not supported with installed version of qulacs. "
+                    "Please install 'qulacs-gpu' to use GPU simulation."
```

Comment:
Nice! Does this plugin depend on `qulacs` or `qulacs-qpu` in `setup.py`? It might useful to use `extra_requires`, so that the user can specify

```console
pip install pennylane-qulacs[gpu]
```

to automatically install the GPU version?

Reply:
Setup uses `qulacs`....

I added 
``    extras_require=[
        'qulacs-gpu',
    ],
``
to setup, was that what you meant?

Reply:
#1 seems to be closely related

Reply:
This also is very relevant: https://github.com/tensorflow/tensorflow/issues/7166

It looks like we have two options:

1. Move both qulacs _and_ qulacs-gpu to `extra_requires`:

  ```python
  setup(
    name="my_package",
    ...,  # other stuff
    install_requires=<list of dependencies EXCLUDING qulacs>,
    extras_require={
        "cpu": ["qulacs>=0.1.10.1"],
        "gpu": ["qulacs-gpu>=0.1.10.1"],
    }
  )
  ```

  This way, users have to do `pip install pennylane-qulacs[cpu]` if they want the CPU version, and `pip install pennylane-qulacs[gpu]` if they want the GPU version. Running `pip install pennylane-qulacs` will pull in all requirements except qulacs.

2. Use logic in `setup.py` to detect if CUDA is installed, and dynamically update `install_requires` accordingly. See https://github.com/ska-sa/montblanc/blob/7a3769dd0a0c192b9c013bbd7fbb181476a0eeab/setup.py#L80 for an example.

I think (1) might be better/easier, we just need to be clear with the installation instructions. While (2) will make a more seamless user experience, I fear it will be hard to maintain and likely could have lots of bugs.

Reply:
I went with option (1) here, and also updated the installation instructions. Let me know if it looks good enough. @josh146

Reply:
Looks perfect @thisac!